Adjust text input scrollWidth and scrollHeight to include padding and any whitespace added by decorations

Ahmad-S792 opened the pull request:
#### 393d4a1bb437c9ee4312816fe53087e53d5fd5ab
<pre>
Adjust text input scrollWidth and scrollHeight to include padding and any whitespace added by decorations

<a href="https://bugs.webkit.org/show_bug.cgi?id=250383">https://bugs.webkit.org/show_bug.cgi?id=250383</a>
rdar://problem/104332108

Reviewed by Alan Baradlay.

This patch is to align WebKit with Blink / Chromium and Gecko / Firefox.

Merge - <a href="https://src.chromium.org/viewvc/blink?view=revision&amp">https://src.chromium.org/viewvc/blink?view=revision&amp</a>;revision=186123

This patch modifies &quot;ScrollWidth&quot; and &quot;ScrollHeight&quot; functions to ensure that
scrollLeft (and similarly scrollTop) is always within range [0, scrollWidth - clientWidth]
and padding and whitespaces added by decorations are accounted.

* Source/WebCore/rendering/RenderTextControlSingleLine.cpp:
(RenderTextConrolSingleLine::scrollWidth):
(RenderTextConrolSingleLine::scrollHeight):
* LayoutTests/fast/scrolling/scroll-max-value.html: Add Test Case
* LayoutTests/fast/scrolling/scroll-max-value-expected.txt: Add Test Case Expectation

Canonical link: <a href="https://commits.webkit.org/261121@main">https://commits.webkit.org/261121@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e240150124d4273ad60da14ee080a849dbecbed1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110605 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19688 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/43187 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1951 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/119461 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/114564 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21101 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10808 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/1477 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116346 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/15711 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/98902 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102912 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/97671 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30569 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/44029 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/12335 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31907 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/85943 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12890 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/8858 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18254 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/51523 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7705 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14826 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderTextControlSingleLine.cpp:
```diff
@@ -366,15 +366,21 @@ void RenderTextControlSingleLine::autoscroll(const IntPoint& position)
 {
-    if (auto innerTextElement = this->innerTextElement(); innerTextElement && innerTextElement->renderer())
-        return innerTextElement->renderer()->scrollWidth();
+    if (auto inner = this->innerTextElement() ? innerTextElement()->renderer() : nullptr) {
```
This is slightly different from what  the blink patch has as you are getting scrollWidth off of the renderer instead of the element, but I guess it's fine.
I would write it like this (there was also a typo in blink's commit.

if (auto* innerTextRenderer = innerTextElement() ? innerTextElement()->renderer() : nullptr) {
   // Adjust scrollWidth to include input element horizontal paddings and decoration width.
   auto adjustment = clientWidth() - innerTextRenderer->clientWidth();
   return innerTextRenderer->scrollWidth() + adjustment;
 }
(same applies to scrollHeight)

Ahmad-S792 replied:
Heading back home. Will update once I have access to personal laptop. 👍

Ahmad-S792 replied:
Done.

alanbaradlay commented on Source/WebCore/rendering/RenderTextControlSingleLine.cpp:
```diff
@@ -366,15 +366,21 @@ void RenderTextControlSingleLine::autoscroll(const IntPoint& position)
+    if (auto* innerTextRenderer = innerTextElement() ? innerTextElement()->renderer() : nullptr) {
+        // Adjust scrollWidth to inculde input element horizontal paddings and decoration width.
+        auto adjustment = clientWidth() - innerTextRenderer->clientWidth();
+        return innerTextElement()->renderer()->scrollWidth() + adjustment;
```
this should read: innerTextRenderer->scrollWidth()

Ahmad-S792 replied:
Ops! My sleepy head.. :-( Fixed now.

alanbaradlay commented on Source/WebCore/rendering/RenderTextControlSingleLine.cpp:
```diff
@@ -366,15 +366,21 @@ void RenderTextControlSingleLine::autoscroll(const IntPoint& position)
+    if (auto* innerTextRenderer = innerTextElement() ? innerTextElement()->renderer() : nullptr) {
+        // Adjust scrollHeight to inculde input element vertical paddings and decoration height.
+        auto adjustment = clientHeight() - innerTextRenderer->clientHeight();
+        return innerTextElement()->renderer()->scrollHeight() + adjustment;
```
this should read: innerTextRenderer-> scrollHeight()

Ahmad-S792 replied:
Fixed.

alanbaradlay reviewed (approved):
thank you!

Ahmad-S792 committed:
Adjust text input scrollWidth and scrollHeight to include padding and any whitespace added by decorations

https://bugs.webkit.org/show_bug.cgi?id=250383
rdar://problem/104332108

Reviewed by Alan Baradlay.

This patch is to align WebKit with Blink / Chromium and Gecko / Firefox.

Merge - https://src.chromium.org/viewvc/blink?view=revision&revision=186123

This patch modifies "ScrollWidth" and "ScrollHeight" functions to ensure that
scrollLeft (and similarly scrollTop) is always within range [0, scrollWidth - clientWidth]
and padding and whitespaces added by decorations are accounted.

* Source/WebCore/rendering/RenderTextControlSingleLine.cpp:
(RenderTextConrolSingleLine::scrollWidth):
(RenderTextConrolSingleLine::scrollHeight):
* LayoutTests/fast/scrolling/scroll-max-value.html: Add Test Case
* LayoutTests/fast/scrolling/scroll-max-value-expected.txt: Add Test Case Expectation

Canonical link: https://commits.webkit.org/261121@main
